File: trees/543_diameter_of_binary_tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def diameter_of_binary_tree(root: Optional[TreeNode]) -> int:
    d = 0

    def dfs(node):
        nonlocal d
        if not node:
            return 0

        left = dfs(node.left)
        right = dfs(node.right)

        d = max(d, left + right)

        return max(left, right) + 1

    dfs(root)
    return d

File: trees/543_diameter_of_binary_tree.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(node *TreeNode, d *int) int {
	if node == nil {
		return 0
	}

	left := maxDepth(node.Left, d)
	right := maxDepth(node.Right, d)
	*d = max(*d, left+right)

	return 1 + max(left, right)
}
func diameterOfBinaryTree(root *TreeNode) int {
	d := 0

	maxDepth(root, &d)
	return d
}